### PR TITLE
ENG-4 add bitbucket

### PR DESCRIPTION
**Pull request by zreigz**
## Summary
Add support for bitbucket

## Checklist

- [ ] If required, I have updated the Plural documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a meaningful title and summary to convey the impact of this PR to a user.
- [ ] I have added relevant labels to this PR to help with categorization for release notes.

### CODE REVIEW

File: pkg/scm/bitbucket.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	if err != nil {
+		return Context{}, err
+	}
+
+	wl, err := b.Client.Workspaces.List()
+	if err != nil {
+		return Context{}, err
+	}
+
+	workspaces := make([]string, 0)
```
Comment (michaeljguarino) [inline pkg/scm/bitbucket.go:80]: you could use polly's map here

File: pkg/scm/bitbucket.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	projectKeys := make(map[string]string, 0)
+	projects := make([]string, 0)
+	for _, p := range pr.Items {
+		projectKeys[p.Name] = p.Key
+		projects = append(projects, p.Name)
+	}
+
+	project := ""
+	if len(projects) == 0 {
+		return Context{}, fmt.Errorf("You don't have any Bitbucket project created. Please create one first \n")
```
Comment (michaeljguarino) [inline pkg/scm/bitbucket.go:113]: reword to `You don't have a BitBucket project created, please create one and try again` perhaps

File: pkg/scm/bitbucket.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		return Context{}, err
+	}
+	sshAddress, err := getSSHAddress(res.Links)
+	if err != nil {
+		return Context{}, err
+	}
+
+	utils.Highlight("Setting up a read-write deploy key for this repo...\n")
+
+	if _, err := b.Client.Repositories.DeployKeys.Create(&bitbucket.DeployKeyOptions{
```
Comment (michaeljguarino) [inline pkg/scm/bitbucket.go:156]: do these have write permissions by default (will be needed)?

File: pkg/scm/bitbucket.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	if err != nil {
+		return Context{}, err
+	}
+	defer os.RemoveAll(dir)
+	gitIgnore := path.Join(dir, ".gitignore")
+	if err := os.WriteFile(gitIgnore, []byte(""), 0644); err != nil {
+		return Context{}, err
+	}
+
+	if err := b.Client.Repositories.Repository.WriteFileBlob(&bitbucket.RepositoryBlobWriteOptions{
```
Comment (michaeljguarino) [inline pkg/scm/bitbucket.go:176]: why do we need to do this?  `plural init` should override the gitignore (i remember another bitbucket user hitting this snag a looong time ago)

File: pkg/scm/bitbucket.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+func (b *Bitbucket) StarPluralGitHubRep() error {
+	return nil
+}
+
+func getSSHAddress(links map[string]interface{}) (string, error) {
+	sshAddress := ""
+	var linkList []Link
+	if links != nil {
+		clone := links["clone"]
```
Comment (michaeljguarino) [inline pkg/scm/bitbucket.go:205]: what's going on here?  Seems very unusual we need to marshal then unmarshal whatever type this is

**Comment by zreigz** [comment]
@michaeljguarino PTAL

File: pkg/scm/bitbucket.go (hunk 5 above)
Comment (michaeljguarino) [reply pkg/scm/bitbucket.go:205]: god, that is bonkers

File: pkg/scm/bitbucket.go (hunk 4 above)
Comment (michaeljguarino) [reply pkg/scm/bitbucket.go:176]: Is there no api that creates the repo with an initial commit?

File: pkg/scm/bitbucket.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+		return Context{}, err
+	}
+	// The bitbucket creates empty repository.
+	// Initialize bitbucket repo with .gitignore file.
+	dir, err := os.MkdirTemp("", "bitbucket")
+	if err != nil {
+		return Context{}, err
+	}
+	defer os.RemoveAll(dir)
+	gitIgnore := path.Join(dir, ".gitignore")
```
Comment (michaeljguarino) [inline pkg/scm/bitbucket.go:171]: Writing a basic README.md would be safer than a gitignore I think (which we do end up generating in `plural init`

File: pkg/scm/bitbucket.go (hunk 4 above)
Comment (michaeljguarino) [reply pkg/scm/bitbucket.go:176]: I feel like i've seen users have an initial commit, maybe it's set at a different level in bitbucket

Comment (zreigz) [reply pkg/scm/bitbucket.go:176]: there is no API for this. Only UI has the option to create README and .gitignore files

File: pkg/scm/bitbucket.go (hunk 6 above)
Comment (zreigz) [reply pkg/scm/bitbucket.go]: done

**Pull request merged**